From e2b11bfe943a30df775f61925d23863f02488eca Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 4 Jun 2023 23:37:21 +0200
Subject: [PATCH 189/196] Add activate for Deuteronomy upgrade (#299)

* Add Joshua upgrade time
* Bump replay protection time
* Add code to activate Deuteronomy upgrade.
---
 src/chainparams.cpp                           |  8 ++
 src/consensus/activation.cpp                  | 10 +++
 src/consensus/activation.h                    |  4 +
 src/consensus/addresses.h                     | 17 ++++
 src/consensus/params.h                        |  4 +
 src/init.cpp                                  |  3 +-
 src/minerfund.cpp                             |  5 ++
 src/validation.cpp                            |  2 +-
 .../logos_feature_minerfund_activation.py     | 79 +++++++++++++++++--
 9 files changed, 123 insertions(+), 9 deletions(-)

diff --git a/src/chainparams.cpp b/src/chainparams.cpp
index 439e52dbb..f8ee3b9d8 100644
--- a/src/chainparams.cpp
+++ b/src/chainparams.cpp
@@ -111,6 +111,8 @@ public:
         consensus.numbersActivationTime = 1671659280;
         // 2023-06-21T14:58:00.000Z protocol upgrade
         consensus.deuteronomyActivationTime = 1687359480;
+        // 2023-12-22T03:27:00.000Z protocol upgrade
+        consensus.joshuaActivationTime = 1703215620;
 
         /**
          * The message start string is designed to be unlikely to occur in
@@ -250,6 +252,10 @@ public:
         consensus.deuteronomyActivationTime =
             mainnetConsensus.deuteronomyActivationTime -
             testnetActivationOffset;
+        // 2023-12-22T03:27:00.000Z protocol upgrade
+        consensus.joshuaActivationTime =
+            mainnetConsensus.joshuaActivationTime -
+            testnetActivationOffset;
 
         // "ltdk" with MSB set
         diskMagic[0] = 0xec;
@@ -360,6 +366,8 @@ public:
         // 2023-06-21T14:58:00.000Z protocol upgrade
         consensus.deuteronomyActivationTime =
             mainnetConsensus.deuteronomyActivationTime;
+        // 2023-12-22T03:27:00.000Z protocol upgrade
+        consensus.joshuaActivationTime = mainnetConsensus.joshuaActivationTime;
 
         // "lrdk" with MSB set
         diskMagic[0] = 0xec;
diff --git a/src/consensus/activation.cpp b/src/consensus/activation.cpp
index 2a7a8fc3a..fa501d269 100644
--- a/src/consensus/activation.cpp
+++ b/src/consensus/activation.cpp
@@ -42,3 +42,13 @@ bool IsNumbersEnabled(const Consensus::Params &params,
 
     return IsNumbersEnabled(params, pindexPrev->GetMedianTimePast());
 }
+
+bool IsDeuteronomyEnabled(const Consensus::Params &params,
+                          const CBlockIndex *pindexPrev) {
+    if (pindexPrev == nullptr) {
+        return false;
+    }
+
+    return pindexPrev->GetMedianTimePast() >=
+           gArgs.GetArg("-deuteronomyactivationtime", params.deuteronomyActivationTime);
+}
diff --git a/src/consensus/activation.h b/src/consensus/activation.h
index ce6b80fb7..4cccc3036 100644
--- a/src/consensus/activation.h
+++ b/src/consensus/activation.h
@@ -27,4 +27,8 @@ bool IsNumbersEnabled(const Consensus::Params &params,
 bool IsNumbersEnabled(const Consensus::Params &params,
                       const CBlockIndex *pindexPrev);
 
+/** Check if June 21st, 2023 protocol upgrade has activated. */
+bool IsDeuteronomyEnabled(const Consensus::Params &params,
+                          const CBlockIndex *pindexPrev);
+
 #endif // BITCOIN_CONSENSUS_ACTIVATION_H
diff --git a/src/consensus/addresses.h b/src/consensus/addresses.h
index 235eadafb..3e1363525 100644
--- a/src/consensus/addresses.h
+++ b/src/consensus/addresses.h
@@ -108,5 +108,22 @@ Consensus::CoinbaseAddresses AddressSets = {
             "lotus_16PSJL7B1pQm4pT3M4k6BeXNR8PXkvuv8BeosgUdA",
             "lotus_16PSJNfvYdAquq6X28omdnpMyixZX9JfF9MgmDX3H",
         },
+    .deuteronomy =
+        {
+            // Note: nobody has private keys for these
+            "lotus_16PSJMzwHUZRNKGRxj6M78GUriT4JE9JVHBJwsPDR",
+            "lotus_16PSJQWWnyZvpXASxfYPnMuZgXuJe9Mi4Ko9E6AJJ",
+            "lotus_16PSJPLfgb7mYuMsyveuCiiw2EefZTDRevVGKrhCz",
+            "lotus_16PSJM5L7t1FvXnXYMKLBw5rBUNDewQwSxAnXnhcC",
+            "lotus_16PSJQaE6rFjErEmzsYaL2dNq51RUn1CPsXFykDeo",
+            "lotus_16PSJMsa5viW6TmE4TpJo2Ln5H5Tqu3bAsfRHEzcz",
+            "lotus_16PSJQkFaM93LHjiZLT2N3sYi8Z5yoPuiCTK6LSFB",
+            "lotus_16PSJM1S7w42U5vfPMHYFj2cuHnhsdVdrdc7BGGBV",
+            "lotus_16PSJQfyDWDvCBayWdWFDkkq8dB1uLfHUc5mt24C5",
+            "lotus_16PSJNYs6oFDXedfTMKBmJanibWk5akJgj6Ly1bxc",
+            "lotus_16PSJNwovGUnVpdWJQDuzX2JkKX14QpRbPVXEf1SZ",
+            "lotus_16PSJQr1FEMMWu6cAW4cWbXSJEkVLVsxwtzoiTKNS",
+            "lotus_16PSJQhqxYBFLYmqkZqvs8cf6ydaaXRxJBCyHCffb",
+        },
 };
 } // namespace RewardAddresses
diff --git a/src/consensus/params.h b/src/consensus/params.h
index f8540d7e7..6ee4c8fe6 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -19,6 +19,7 @@ struct CoinbaseAddresses {
     std::vector<std::string> exodus;
     std::vector<std::string> leviticus;
     std::vector<std::string> numbers;
+    std::vector<std::string> deuteronomy;
 };
 
 /**
@@ -39,6 +40,9 @@ struct Params {
     /** Unix time used for MTP activation of 2023-06-21T14:58:00.000Z protocol
      * upgrade */
     int deuteronomyActivationTime;
+    /** Unix time used for MTP activation of 2023-12-22T03:27:00.000Z protocol
+     * upgrade */
+    int joshuaActivationTime;
 
     /**
      * Don't warn about unknown BIP 9 activations below this height.
diff --git a/src/init.cpp b/src/init.cpp
index 60de7eecc..4c2549ad6 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -424,7 +424,8 @@ void SetupServerArgs(NodeContext &node) {
         "-min", "-resetguisettings", "-rootcertificates=<file>", "-splash",
         "-uiplatform",
         // TODO remove after the December 2021 upgrade
-        "-exodusactivationtime", "-leviticusactivationtime", "-numbersactivationtime"};
+        "-exodusactivationtime", "-leviticusactivationtime",
+        "-numbersactivationtime", "-deuteronomyactivationtime"};
 
     // Set all of the args and their help
     // When adding new options to the categories, please keep and ensure
diff --git a/src/minerfund.cpp b/src/minerfund.cpp
index 5bd614625..eab1aab40 100644
--- a/src/minerfund.cpp
+++ b/src/minerfund.cpp
@@ -56,6 +56,11 @@ std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
         return {};
     }
 
+    if (IsDeuteronomyEnabled(params, pindexPrev)) {
+        return BuildOutputsCycling(params.coinbasePayoutAddresses.deuteronomy,
+                                   pindexPrev, blockReward);
+    }
+
     if (IsNumbersEnabled(params, pindexPrev)) {
         return BuildOutputsCycling(params.coinbasePayoutAddresses.numbers,
                                    pindexPrev, blockReward);
diff --git a/src/validation.cpp b/src/validation.cpp
index e274bf8d3..fb51c77f4 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -278,7 +278,7 @@ bool CheckSequenceLocks(const CTxMemPool &pool, const CTransaction &tx,
 static bool IsReplayProtectionEnabled(const Consensus::Params &params,
                                       int64_t nMedianTimePast) {
     return nMedianTimePast >= gArgs.GetArg("-replayprotectionactivationtime",
-                                           params.deuteronomyActivationTime);
+                                           params.joshuaActivationTime);
 }
 
 static bool IsReplayProtectionEnabled(const Consensus::Params &params,
diff --git a/test/functional/logos_feature_minerfund_activation.py b/test/functional/logos_feature_minerfund_activation.py
index f5fa36caa..c8ee7d60b 100755
--- a/test/functional/logos_feature_minerfund_activation.py
+++ b/test/functional/logos_feature_minerfund_activation.py
@@ -28,6 +28,10 @@ from test_framework.util import assert_equal
 EXODUS_ACTIVATION_TIME = 2000000000
 LEVITICUS_ACTIVATION_TIME = 2010000000
 NUMBERS_ACTIVATION_TIME = 2020000000
+DEUTERONOMY_ACTIVATION_TIME = 2030000000
+JOSHUA_ACTIVATION_TIME = 2040000000
+
+REPLAYPROTECTION_ACTIVATION_TIME = JOSHUA_ACTIVATION_TIME
 
 # see consensus/addresses.h, use getaddressinfo to get the scriptPubKey
 GENESIS_SCRIPTS = [
@@ -90,6 +94,21 @@ NUMBERS_SCRIPTS = [
     "76a91461cebabeaa6365887bb95d95e62b848fa365506388ac",
     "76a914b59008144c4f883e3b1fd9e18e479bc2f877fca788ac",
 ]
+DEUTERONOMY_SCRIPTS = [
+    "76a9149f9c4d8a7b187bda082aaa151f504284210f937c88ac",
+    "76a914f19444ce2d31e7cc825123dd695bc3f0d5aa8d7188ac",
+    "76a914cb60a8cacd4fefc5843d30c57af8a097d3092c0b88ac",
+    "76a914816d58bec68acb2df40ed1edb791af22d8741b1e88ac",
+    "76a914f3ab5b4edc3d4c7a8f4036c8874eb1924325dd1688ac",
+    "76a9149b769ea034cd3bba1349c78d045806a4657d3fac88ac",
+    "76a914f950726f65f10bf3ec7f1bbd236a1254769608be88ac",
+    "76a9147f3bb0818f6f715b9609c0bf624d8741dbadce2f88ac",
+    "76a914f6e7383b4c2e3caf6a0360f04fb2274195ca9c0488ac",
+    "76a914b1967870fea5b1cf9b197841965758aa5aebc4cf88ac",
+    "76a914be81d86c6caa6fea587b5e9d26ddffe662dd3bb888ac",
+    "76a914fc8db0e10f08058d707c36ec57bd1d2f88fee47c88ac",
+    "76a914f7f5779b38137107294ab6f75adf5f2b94fc25b688ac",
+]
 
 
 class MinerFundActivationTest(BitcoinTestFramework):
@@ -101,15 +120,19 @@ class MinerFundActivationTest(BitcoinTestFramework):
             f'-exodusactivationtime={EXODUS_ACTIVATION_TIME}',
             f'-leviticusactivationtime={LEVITICUS_ACTIVATION_TIME}',
             f'-numbersactivationtime={NUMBERS_ACTIVATION_TIME}',
+            f'-deuteronomyactivationtime={DEUTERONOMY_ACTIVATION_TIME}',
+            f'-replayprotectionactivationtime={REPLAYPROTECTION_ACTIVATION_TIME}',
         ]]
 
     def run_test(self):
         node = self.nodes[0]
         node.add_p2p_connection(P2PDataStore())
 
-        # Mine a block, contains coinbase with miner fund
-        anyonecanspend_address = node.decodescript('51')['p2sh']
-        node.generatetoaddress(1, anyonecanspend_address)
+        # Mine a block, use coin to test replay protection
+        key = node.get_deterministic_priv_key()
+        key_script = node.validateaddress(key.address)['scriptPubKey']
+        coinblockhash = node.generatetoaddress(1, key.address)[0]
+        cointxid = node.getblock(coinblockhash)['tx'][0]
 
         def make_block_with_cb_scripts(scripts):
             best_block_hash = node.getbestblockhash()
@@ -212,10 +235,10 @@ class MinerFundActivationTest(BitcoinTestFramework):
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), None)
 
-        # Set mocktime for Leviticus activation
+        # Set mocktime for Numbers activation
         node.setmocktime(NUMBERS_ACTIVATION_TIME)
-        # Mine 11 blocks with LEVITICUS_ACTIVATION_TIME in the middle
-        # That moves MTP exactly to LEVITICUS_ACTIVATION_TIME
+        # Mine 11 blocks with NUMBERS_ACTIVATION_TIME in the middle
+        # That moves MTP exactly to NUMBERS_ACTIVATION_TIME
         for i in range(-6, 6):
             block = make_block_with_cb_scripts(LEVITICUS_SCRIPTS)
             block.nTime = NUMBERS_ACTIVATION_TIME + i
@@ -233,7 +256,7 @@ class MinerFundActivationTest(BitcoinTestFramework):
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
 
-        # Using new scripts now works
+        # Now the using the leviticus addresses fails
         block = make_block_with_cb_scripts(LEVITICUS_SCRIPTS)
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
@@ -244,6 +267,48 @@ class MinerFundActivationTest(BitcoinTestFramework):
             prepare_block(block)
             assert_equal(node.submitblock(ToHex(block)), None)
 
+        # Using the deuteronomy addresses before upgrade fails
+        block = make_block_cb_post_numbers(DEUTERONOMY_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # Set mocktime for Deuteronomy activation
+        node.setmocktime(DEUTERONOMY_ACTIVATION_TIME)
+        # Mine 11 blocks with DEUTERONOMY_ACTIVATION_TIME in the middle
+        # That moves MTP exactly to DEUTERONOMY_ACTIVATION_TIME
+        for i in range(-6, 6):
+            block = make_block_cb_post_numbers(NUMBERS_SCRIPTS)
+            block.nTime = DEUTERONOMY_ACTIVATION_TIME + i
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
+        assert_equal(node.getblockchaininfo()['mediantime'],
+                     DEUTERONOMY_ACTIVATION_TIME)
+        
+        # Now the using the genesis, exodus, leviticus, or numbers addresses fails
+        for block in [
+            make_block_with_cb_scripts(GENESIS_SCRIPTS),
+            make_block_with_cb_scripts(EXODUS_SCRIPTS),
+            make_block_with_cb_scripts(LEVITICUS_SCRIPTS),
+            make_block_cb_post_numbers(NUMBERS_SCRIPTS),
+        ]:
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+
+        # Using the Deuteronomy scripts now works
+        for i in range(0, 26):
+            block = make_block_cb_post_numbers(DEUTERONOMY_SCRIPTS)
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
+        # Check replay protection is not enabled yet
+        tx = CTransaction()
+        tx.vin = [CTxIn(COutPoint(int(cointxid, 16), 1))]
+        tx.vout = [CTxOut(int(SUBSIDY * COIN) // 2 - 1000,
+                          CScript.fromhex(key_script))]
+        raw_tx = node.signrawtransactionwithkey(tx.serialize().hex(), [key.key])
+        node.sendrawtransaction(raw_tx['hex'])
+
 
 if __name__ == '__main__':
     MinerFundActivationTest().main()
-- 
2.39.5 (Apple Git-154)

